src/documents/posts/2016-06-06-sticky-ab-tests-with-fastly.md

```diff
@@ -0,0 +1,91 @@
+
+<!-- more -->
+
+We had some very promising results from some early Apache Bench runs, but how might we determine the real-world performance difference between the two platforms?  Futhermore, how might we determine if there is a difference in user conversion?  For those that haven't heard, [Loading Time Affects Your Bottom Line](https://blog.kissmetrics.com/loading-time/).
+
+We decided the answer to both would be to set up an A/B test with our website visitors.  We'd route half of our customers to Heroku and half to Ranch, and then analyize the results.  To measure conversion over time, we'd make the choice once per browser rather than once per request, aka "sticky."  (Once per customer would be ideal, but not worth the extra effort.)
+
+Luckily, we already used Fastly for SSL termination and edge caching.  What some folks don't know is that you can also upload custom Varnish Config Language (VCL) files to Fastly.  Starting with [an example from Fastly's docs](https://www.fastly.com/blog/best-practices-for-using-the-vary-header), I came up with this VCL:
+
+```vcl
+
+sub vcl_recv {
+#FASTLY recv
+
+  if (req.http.Cookie ~ "platform=") {
+    // TBD
+  } else {
+    // 50% to Ranch
+    if (randombool(50, 100)) {
+      set req.http.GE-Set-Platform = "ranch";
+    } else {
+      set req.http.GE-Set-Platform = "heroku";
+    }
+  }
+
+}
+
+sub vcl_deliver {
+#FASTLY deliver
+
+  if (req.http.GE-Set-Platform) {
+    add resp.http.Set-Cookie = "platform=" req.http.GE-Set-Platform "; Domain=goodeggs.com; Path=/; Expires=" now + 365d ";";
+  }
+
+}
+
+[...snip...]
+```
+
+This picks a platform once per browser (per 365 days), and store the choice in a cookie called `platform`.  Unlike Fastly's example, we won't set the cookie on every response.  (I should also note I'm not concerned with any `Vary` header manipulation because my platforms serve the same content.  Be careful -- YMMV.)
+
+But so far, nothing happens.  Let's fix that and switch backends based on our choice:
+
+```vcl
+[...snip...]
+
+backend F_heroku { ... }
+backend F_ranch { ... }
+
+sub vcl_recv {
+#FASTLY recv
+
+  if (req.http.Cookie ~ "platform=ranch") {
+    set req.backend = F_ranch;
+  else if (req.http.Cookie ~ "platform=heroku") {
+    set req.backend = F_heroku;
+  } else {
+    // 50% to Ranch
+    if (randombool(50, 100)) {
+      set req.http.GE-Set-Platform = "ranch";
+      set req.backend = F_ranch;
+    } else {
+      set req.http.GE-Set-Platform = "heroku";
+      set req.backend = F_heroku;
+    }
+  }
+
+}
+
+[...snip...]
+
+```
+
+And with that, we're routing traffic to both platforms.
+
+Of course there's the problem of reporting and analyzing the data, but I'll leave that as an excercise to the reader.  In our case, our application log lines include response time, and we were able to tease apart the log lines and do the analysis in Sumo Logic.
```

Comment:
this is probably the first of a few related blog posts, one of which will
be a deep-dive on the numbers co-authored with Alon.  felt like I'd just
keep this technical and skip the results.  @alon, thoughts?


On Mon, Jun 6, 2016 at 3:16 PM, Lei <notifications@github.com> wrote:

> In src/documents/posts/2016-06-06-sticky-ab-tests-with-fastly.md
> <https://github.com/goodeggs/bites/pull/57#discussion_r65979810>:
>
> > +      set req.backend = F_ranch;
> > +    } else {
> > +      set req.http.GE-Set-Platform = "heroku";
> > +      set req.backend = F_heroku;
> > +    }
> > +  }
> > +
> > +}
> > +
> > +[...snip...]
> > +
> > +```
> > +
> > +And with that, we're routing traffic to both platforms.
> > +
> > +Of course there's the problem of reporting and analyzing the data, but I'll leave that as an excercise to the reader.  In our case, our application log lines include response time, and we were able to tease apart the log lines and do the analysis in Sumo Logic.
>
> hah!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/goodeggs/bites/pull/57/files/3626dde50701e5d9bab0b8d58ff9f597be8f5ffa#r65979810>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABB0nfVt1ULas5cSTp-gUva3kFFvwXjks5qJJxYgaJpZM4IvQAs>
> .
>